velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -105,12 +108,269 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+            ::substrait::Expression_Literal::LiteralTypeCase::kList) {
+      getOffsetsAndSizes(offsets, sizes, lengths[level]);
+      // use flatVector as elements for innermost arrayVector
+      return std::make_shared<ArrayVector>(
+          pool_,
+          ARRAY(flatVector->type()),
+          nullptr,
+          listLiteral.list().values().size(),
+          offsets,
+          sizes,
+          flatVector);
+    }
+    VectorPtr elements = constructArray(
+        listLiteral.list().values(0), flatVector, level + 1, lengths);
+    getOffsetsAndSizes(offsets, sizes, lengths[level + 1]);
+    return std::make_shared<ArrayVector>(
+        pool_,
+        ARRAY(elements->type()),
+        nullptr,
+        listLiteral.list().values().size(),
+        offsets,
+        sizes,
+        elements);
+  }
+  VELOX_FAIL("typeCase must be kList.");
+}
+
+template <TypeKind kind>
+void listLiteralToFlatVector(
+    const ::substrait::Expression::Literal& listLiteral,
+    const VectorPtr& vector,
+    vector_size_t& index) {
+  using TFlat = typename KindToFlatVector<kind>::type;
+  using TCpp = typename TypeTraits<kind>::NativeType;
+  auto flatVector = vector->as<TFlat>();
+  auto typeCase = listLiteral.literal_type_case();
+  if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kList) {
+    if (listLiteral.list().values().size() == 0) {
+      return;
+    }
+    for (auto child : listLiteral.list().values()) {
+      listLiteralToFlatVector<kind>(child, vector, index);
+    }
+    return;
+  }
+
+  if constexpr (std::is_same_v<TCpp, StringView>) {
+    if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kString) {
+      flatVector->set(index++, StringView(listLiteral.string()));
+    }
+    if (typeCase ==
+        ::substrait::Expression_Literal::LiteralTypeCase::kVarChar) {
+      flatVector->set(index++, StringView(listLiteral.var_char().value()));
+    }
+  } else if constexpr (
+      std::is_same_v<TCpp, Timestamp> || std::is_same_v<TCpp, Date> ||
+      std::is_same_v<TCpp, IntervalDayTime>) {
+  } else {
+    if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kNull) {
+      vector->setNull(index++, true);
+    } else {
+      flatVector->set(index++, GetLiteralValue<TCpp>(listLiteral));
+    }
+  }
+  return;
+}
+
+VectorPtr SubstraitVeloxExprConverter::constructFlatVector(
+    const ::substrait::Expression::Literal& listLiteral,
+    const vector_size_t size,
+    const TypePtr& type) {
+  VELOX_CHECK(type->isPrimitiveType());
+  auto vector = BaseVector::create(type, size, pool_);
+
+  vector_size_t index = 0;
+  auto kind = vector->typeKind();
+  VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      listLiteralToFlatVector, kind, listLiteral, vector, index);
+  return vector;
+}
+
+ArrayVectorPtr SubstraitVeloxExprConverter::literalsToArrayVector(
+    const ::substrait::Expression::Literal& listLiteral) {
+  if (listLiteral.list().values().size() == 0) {
+    std::vector<variant> variants;
```

Comment:
@zhejiangxiaomai 

This code doesn't look right. It is not clear what would be the type of the ArrayVector. It also seems to me this code path is not covered by tests. Would you take a look?

Reply:
OK, This is for array[].

Reply:
@zhejiangxiaomai Actually, I see the test coverage now. 

Let's not use variants, but create ArrayVector directly.

```
    // Create empty array vector.
    BufferPtr offsets = allocateOffsets(1, pool_);
    BufferPtr sizes = allocateOffsets(1, pool_);
    return std::make_shared<ArrayVector>(
        pool_, ARRAY(UNKNOWN()), nullptr, 1, offsets, sizes, nullptr);
```

Reply:
Thanks, will fix it.